### ISSUE TITLE
Suggestions for micro/mini-optimizations

### ISSUE THREAD

**Issue by dsuch** [benoitc/gunicorn#1612]
Hello,

I was checking out the flow of data from GeventWorker.handle_request down to the WSGI application configured for a gunicorn-based server in order to see if anything can be made more efficient.

What I'm writing about below are comments for a future prospective contributor to pick up.

I won't be able to implement it all myself because after several years of rock-solid stability with gunicorn I will likely need to start to use another server - the reason being that I need and will need to support Python 2.7 for years ahead which will not be possible after #1288 and related tickets *(update: as noted by @berkerpeksag in a comment below, even after gunicorn 20 has been released, there will be still v. 19 with support for Python 2.7)*.

Most of the changes are trivial and insignificant in isolation but they add up and I'm in a situation where I'm improving the performance of [my application](https://github.com/zatosource/zato) bit-by-bit, literally going up 1%-3% each week or month, benchmarking and profiling the code on a regular basis and gunicorn is part of it.

Basically, what I've been doing is finding things that don't need to be (re-)computed or looked up on each request. Micro- or at least mini-optimizations, in other words.

They can be broken down into several categories:

1. **Eliminating object name look-ups.**

This is not particularly beautiful but in the hottest paths this can be employed. For instance here https://github.com/benoitc/gunicorn/blob/d0d1a1ca99ea538bf68186b17e891b44abae7a98/gunicorn/http/parser.py#L14 - ```def __init__(self, cfg, source):``` can be changed to ```def __init__(self, cfg, source, _SocketUnreader=SocketUnreader, _IterUnreader=IterUnreader):``` and then the already resolved local names can be used in this method. 

The same can be done in a dozen or more places on the core message parsing path. E.g. here https://github.com/benoitc/gunicorn/blob/master/gunicorn/workers/async.py#L34 - in http.RequestParser first 'http' and then its attribute 'RequestParser' is looked up on each request even if they are not going to change. Same goes for six.next, this will not change. http.errors.NoMoreData and all the other objects can be made local. There are dozens of places like it.

Like I said, this is not something pleasant, I'm not advocating this type of programming and I'm certainly not using it myself anywhere except for the hottest path that in a code-base of mine takes up ninety-nine percent of processing time even if in terms of lines of code lines it's only one percent.

2. **Not using the logger if it's not needed**

This is not normally seen but, once the heavier pieces are lifted, it turns out that the stdlib's logger adds some overhead. I remember a seemingly trivial change in gevent in one of 1.0 beta versions that suddenly added an access logger which, even if not used, contributed at least 10% of overhead to each request. A similar case, though not that prominent, is here with gunicorn where self.log.access is used even if the underlying WSGI application (the actual server) may very well use its own logging which is the case in my application so self.log.access is unnecessary. I would suggest adding a flag to configuration, a boolean logging enabled/disabled so as not to call self.log.access at all (unless if self.log: that is).

Again, this kind of 'if' check makes the code-base less attractive.

Actually, in Zato we even monkey-patch (yes) stdlib's logging.Logger._log to get some small, but measurable, performance boost https://github.com/zatosource/zato/blob/main/code/zato-common/src/zato/common/microopt.py

3. **Not consulting .cfg in each request**

This is similar to bullet no. 1 - for instance here https://github.com/benoitc/gunicorn/blob/master/gunicorn/http/message.py#L38 - self.limit_request_fields and similar are computed in each request but there is no need to, they won't change. They could be pre-computed once. There are a few more places like this one.

On a related note, MAX_HEADERS could be likely changed to 100 or even 50, probably?

4. **Moving some logic to parse_headers**

Over here https://github.com/benoitc/gunicorn/blob/master/gunicorn/http/message.py#L59 all headers are parsed and this could be a good place to set things such as a new self.should_close flag while the headers are being iterated over instead of having to call should_close() on each request, which also needs to iterate over the headers (without parsing, for sure, but this is a loop that could be avoided). The headers list is referred to in a few other places - in the same module + when building the WSGI environment.

5. **Pre-compiling regular expressions**

There used to be a point about the fact that header and version regular expressions are compiled in each request but I'm on gunicorn 18 in Zato and I can see that it was already done in a newer version of gunicorn :-) 

But, there is another regexp module called https://pypi.python.org/pypi/regex which could be faster than stdlib's re. I know it is faster in some places in Zato but whether it would help in gunicorn, I'm not sure.

6. **Making optional calls no-ops**

For instance, self.cfg.pre_request and cfg.post_request are called in each request and they log request method + path on DEBUG level which is something that could be turned off completely if someone had their own logger. Again, something like 'if self.cfg.pre_request: self.cfg.pre_request(...)' would be needed (but not if hasattr(self.cfg, 'pre_request')).

A similar situation is here https://github.com/benoitc/gunicorn/blob/master/gunicorn/http/message.py#L169 - the call to self.proxy_protocol is needed only if self.cfg.proxy_protocol was enabled so it could be made before the call even if that method returns immediately.

So basically, this all can be summarized to:

* Let's pre-compute on a process level things that won't change unless a process restarts
* Let's minimize method calls
* Let's minimize namespace look-ups
* Possibly, check out faster external libraries (e.g. regex instead of re)

I haven't looked into the whole algorithm/logic of parsing so these are only simple micro/mini-optimizations. But I just wanted to share my findings, perhaps someone will find them useful and on a lonely afternoon will give it a go.

I have added the changes above to various places and the difference was 3%-10% in terms of req/s - depending on what I was replacing and to what extent - it certainly isn't the case that changing one thing in one method made any difference, but doing it in 10 places made sense.

And again, I'm just reporting what can be done if one is in a situation like myself which is a systematic process of thinning everything down on the critical path.

Also, I haven't checked how responses are built but the same could possibly apply.

Regards.

**Comment by dsuch** [comment]
Let me add that I'm not using PyPy so I cannot comment how important this all would be with a JIT. Hopefully, not at all.

I was testing it with Python 2.7.6 on Ubuntu 14.04 64-bit.

**Comment by tilgovi** [comment]
Thanks, @dsuch. This detail is all very useful.

**Comment by jamadden** [comment]
I can also comment that I will be supporting Python 2.7 for the foreseeable future.

With regards to PyPy I can add some minor points:

- I believe PyPy has also announced support for Python 2.7 for the foreseeable future, although I'm not finding a link right now, but PyPy itself is written in RPython, and as far as I know is still only Python 2 compatible
- Global variables [are not a problem in PyPy](http://pypy.org/performance.html#global-variables). They're basically optimised away if they don't change. "Inlining" them might actually be a small hit since they increase code size.
- The `regex` library is implemented in C, and while it does work on PyPy (I helped make sure of that), any C extension on PyPy is (currently) bound to be slower than a pure-Python implementation. There is currently talk on python-dev about improving performance of `re` on CPython (mostly load/compile time).

I'm curious to see what results the OP would get by testing the application under PyPy2, assuming it's not majorly dependent on native libs that aren't CFFI.

**Comment by berkerpeksag** [comment]
> I won't be able to implement it all myself because after several years of rock-solid stability with gunicorn I will likely need to start to use another server - the reason being that I need and will need to support Python 2.x for years ahead which will not be possible after #1288 and related tickets.

Just an FYI, we are going to maintain 19.x (with Python 2 support) even after 20.x (Python 3 only) is released.

**Comment by dsuch** [comment]
@berkerpeksag - thanks for the update, I appreciate that and it's good to know it.

Thanks @jamadden, yes, I'm curious about it myself. PyPy was not able to build some of our dependencies about a year and a half ago or so but I definitely keep it in mind for the future.

**Comment by jamadden** [comment]
@dsuch PyPy has come a *long* way in terms of CPython C extension API compatibility in the last year and a half and most things will build now. Of course, if you use the C extensions heavily, you may not have good performance, depending on what they do.

**Comment by benoitc** [comment]
no activity since awhile , i am closing the issue. Thanks for the tips and discussion. Some of the tips have been used in recent code. I guess the best way to continue the work now is via some PR. Thanks again :)
